### ISSUE TITLE
[Feature Request] Added the support of evals for the `train` function for early stopping

### ISSUE THREAD

**Issue by DigitalPig** [dask/dask-xgboost#38]
Hi,

Currently the dask-xgboost package [does not](https://github.com/dask/dask-xgboost/blob/4661c8a1d3a7f6b63ff994b944b6a6231e7c9f31/dask_xgboost/core.py#L313-L315) support early_stopping.

Not super familiar with dask or xgboost, but my very rough thinking is it should be just passing the validation set `DMatrix` object during the training function [here](https://github.com/dask/dask-xgboost/blob/4661c8a1d3a7f6b63ff994b944b6a6231e7c9f31/dask_xgboost/core.py#L83-L90)? And the way of finding the location of validation set on each worker should be similar to [what we are doing now with training dataset](https://github.com/dask/dask-xgboost/blob/4661c8a1d3a7f6b63ff994b944b6a6231e7c9f31/dask_xgboost/core.py#L111-L135)?

Please let me know if I am on the right track. I would love to see if I can get this eval functions to work for dask-xgboost but would like to avoid falling into apparent traps as I am not familiar with the backend xgboost/dask too much. Thank you!

**Comment by mrocklin** [comment]
Hi @DigitalPig , 

This package just sets up XGBoost and hands it data and keyword arguments.  The normal XGBoost library then takes control and handles everything. 

So so for features like early stopping you should probably ask in the XGBoost issue tracker rather than here.  We can't control features like this.

**Comment by DigitalPig** [comment]
Hi @mrocklin , Thank you for the reply! Totally understand the "wrapper" function of `dask-xgboost` here. The reason I am asking is because xgboost's documentation says its train can handle early stopping as long as you pass a DMatrix object into it. That is the reason I am asking here to see if there is any specific reason why it is not implemented here. Because if you already thought about it before and had a significant technical issue then I would stop at here too. :)

**Comment by TomAugspurger** [comment]
I don't recall, sorry.

**Comment by mrocklin** [comment]
I wonder if this is handled by the get_xgb_params line just above?

**Comment by TomAugspurger** [comment]
It's a break with convention (but we could probably mirror the break, I'm
not sure).

Something like `sample_weights` has to be passed here since it's data
dependent.

I think supporting `early_stopping_rounds` requires supporting `eval_set`.
I'm not sure if there
are potential issues there with `eval_set` being a distributed dataframe or
array.

**Comment by mmccarty** [comment]
I've got an initial PR for this that I will push up for feedback as soon as I can.
